rust/core-lib/src/find.rs

```diff
@@ -0,0 +1,256 @@
+// Copyright 2018 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Module for searching text.
+
+use std::cmp::{min,max};
+
+use serde_json::Value;
+
+use index_set::IndexSet;
+use xi_rope::delta::{Delta, DeltaRegion};
+use xi_rope::find::{find, CaseMatching};
+use xi_rope::rope::{Rope, LinesMetric, RopeInfo};
+use xi_rope::tree::Cursor;
+use xi_rope::interval::Interval;
+use selection::{Selection, SelRegion};
+use xi_rope::tree::Metric;
+
+/// Contains logic to search text
+pub struct Find {
+    // todo: link to search query so that search results can be correlated back to query
+
+    /// The occurrences, which determine the highlights, have been updated.
+    hls_dirty: bool,
+    /// The currently active search string
+    search_string: Option<String>,
+    /// The case matching setting for the currently active search
+    case_matching: CaseMatching,
+    /// The set of all known find occurrences (highlights)
+    occurrences: Option<Selection>,
+    /// Set of ranges that have already been searched for the currently active search string
+    valid_search: IndexSet,
+}
+
+impl Find {
+    pub fn new() -> Find {
```

Comment:
This is a place where you might consider the `Default` trait, which I personally prefer anytime my constructor has no arguments. Definitely not a big deal though, and probably comes down to preference.